sdk/core/azure-core/src/base64.cpp

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/base64.hpp"
+#include "azure/core/platform.hpp"
+
+#if defined(AZ_PLATFORM_WINDOWS)
+#include <windows.h>
+#elif defined(AZ_PLATFORM_POSIX)
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#endif
+
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Core {
+
+#if defined(AZ_PLATFORM_WINDOWS)
+
+  std::string Base64Encode(const std::vector<uint8_t>& data)
+  {
+    std::string encoded;
+    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4
+    DWORD encodedLength = static_cast<DWORD>((data.size() + 2) / 3 * 4);
+    encoded.resize(encodedLength);
+
+    CryptBinaryToStringA(
+        reinterpret_cast<const BYTE*>(data.data()),
+        static_cast<DWORD>(data.size()),
+        CRYPT_STRING_BASE64 | CRYPT_STRING_NOCRLF,
+        static_cast<LPSTR>(&encoded[0]),
+        &encodedLength);
+
+    return encoded;
+  }
+
+  std::vector<uint8_t> Base64Decode(const std::string& text)
+  {
+    std::vector<uint8_t> decoded;
+    // According to RFC 4648, the encoded length should be ceiling(n / 3) * 4, so we can infer an
+    // upper bound here
+    DWORD decodedLength = DWORD(text.length() / 4 * 3);
+    decoded.resize(decodedLength);
+
+    CryptStringToBinaryA(
+        text.data(),
+        static_cast<DWORD>(text.length()),
+        CRYPT_STRING_BASE64 | CRYPT_STRING_STRICT,
+        reinterpret_cast<BYTE*>(decoded.data()),
+        &decodedLength,
+        nullptr,
+        nullptr);
+    decoded.resize(decodedLength);
+    return decoded;
+  }
+
+#elif defined(AZ_PLATFORM_POSIX)
+
+  std::string Base64Encode(const std::vector<uint8_t>& data)
+  {
+    BIO* bio = BIO_new(BIO_s_mem());
```

Comment:
I think it would be safer to wrap this object into a class where we can have a destructor that calls `BIOS_free_all`

for instance
```cpp
class AzureBio {
private:
   BIO * bio;
public:
   AzureBio() : bio(BIO_new() {
   }
   ~AzureBio() {
     BIO_free_all(bio);
}

// Then, we can safely create a unique_ptr like
auto bio = std::make_unique<BIO_push>();
// Now we don't have to worry about calling free at the end or when something throw. Our wrapper will take care.
```